notebooks/manifold_learning.livemd

```diff
@@ -0,0 +1,155 @@
+
+```elixir
+{tensor_data, key} = EXLA.jit(&Nx.Random.shuffle(&1, &2)).(Nx.Random.key(42), tensor_data)
+
+trimap_res =
+  EXLA.jit(
+    &Scholar.Manifold.Trimap.embed(&1,
+      key: &2,
+      num_components: 2,
+      num_inliers: 12,
+      num_outliers: 4,
+      weight_temp: 0.5,
+      learning_rate: 0.1,
+      metric: :squared_euclidean
+    )
+  ).(
+    tensor_data[[0..10000, ..]],
+    Nx.Random.key(55)
+  )
+```
+
+Now, lets plot the results of Trimap algorithm
+
+```elixir
+coords = [
+  x: trimap_res[[.., 0]] |> Nx.to_flat_list(),
+  y: trimap_res[[.., 1]] |> Nx.to_flat_list()
+]
+
+Tucan.layers([
+  Tucan.scatter(coords, "x", "y", point_size: 1)
+])
+|> Tucan.set_size(300, 300)
+|> Tucan.set_title(
+  "Mammoth dataset with reduced dimensionality using Trimap",
+  offset: 25
+)
+```
+
+For sure, we can recognize mammoth on this picture. Trimap indeed preserved the global structure of data. Result is similar to the projection of 3D mammoth to the YZ plane. Now, plot this projection and compare these two plots.
+
+```elixir
+coords = [
+  x: tensor_data[[0..10000, 1]] |> Nx.to_flat_list(),
+  y: tensor_data[[0..10000, 2]] |> Nx.to_flat_list()
+]
+
+Tucan.layers([
+  Tucan.scatter(coords, "x", "y", point_size: 1)
+])
+|> Tucan.set_size(300, 300)
+|> Tucan.set_title(
+  "Mammoth data set with reduced dimensionality using trimap",
+  offset: 25
+)
+```
+
+These two plots are similiar but there are some important differences. Even if the second figure seems "prettier" it is less informative than the result of trimap. On the first figure, we can spot two tusks while one the second one they overlap and we see only one. Similarly, legs overlay on the first plot and one the second one they are spread and don't intersect with each other.
+
+Now, lets try different algorithm: t-SNE
+
+```elixir
+tsne_res =
+  EXLA.jit(
+    &Scholar.Manifold.TSNE.fit(&1,
+      key: &2,
+      num_components: 2,
+      perplexity: 125,
+      exaggeration: 10.0,
+      learning_rate: 500,
+      metric: :squared_euclidean
+    )
+  ).(
+    tensor_data[[0..2000, ..]],
+    Nx.Random.key(55)
+  )
+```
+
+```elixir
+coords = [
+  x: tsne_res[[.., 0]] |> Nx.to_flat_list(),
+  y: tsne_res[[.., 1]] |> Nx.to_flat_list()
+]
+
+Tucan.layers([
+  Tucan.scatter(coords, "x", "y", point_size: 1)
+])
+|> Tucan.set_size(300, 300)
+|> Tucan.set_title(
+  "Mammoth dataset with reduced dimensionality using Trimap",
+  offset: 25
+)
+```
+
+As we see, t-SNE have completely different results than trimap. This is because t-SNE has a completely different mathematical background of computation. Also t-SNE is slower algorithm, so it can't be used to such big datasets as trimap. However, t-SNE still preserve some features of mammoth like small tusks, feets, and corp. You can experiment with parameter *perplexity* which can significantly change the output of the algorithm.
```

Comment:
```suggestion
As we see, t-SNE gives completely different results than trimap. This is because t-SNE has a completely different mathematical background of computation. Also t-SNE is slower algorithm, so it can't be used on such big datasets as trimap. However, t-SNE preserves some features of mammoth like small tusks, feets, and corp. You can experiment with parameter *perplexity* which can substantially change the output of the algorithm.
```